scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py

```diff
@@ -0,0 +1,263 @@
+            raise TwisterHarnessException(msg)
+        return data
+
+    def readlines(self) -> list[str | object]:
+        """
+        Read all available output lines produced by device from internal buffer.
+        """
+        # give a moment for reader thread to collect output lines
+        time.sleep(0.1)
+        lines = []
+        while not self._device_read_queue.empty():
+            line = self.readline(0.1, raise_timeout_exc=False)
+            lines.append(line)
+        return lines
+
+    def flush(self) -> None:
+        """
+        Remove all available output produced by device from internal buffer.
+        """
+        self.readlines()
+
+    def write(self, data: bytes) -> None:
+        """Write data bytes to device."""
+        if not self._is_device_connected():
+            msg = 'No connection to the device'
+            logger.error(msg)
+            raise TwisterHarnessException(msg)
+        self._write_to_device(data)
+
+    def _initialize_log_files(self) -> None:
+        """
+        TODO:
+        Do we need this?
+        Maybe we should add to log file info about currently executed test?
+        What about multiple devices used in the same test?
+        """
+        log_file_paths = [self.handler_log_path, self.device_log_path]
+        for log_file_path in log_file_paths:
+            with open(log_file_path, 'a+') as log_file:
+                log_file.write(f'\n==== Logging started at {datetime.now()} ====\n')
+
+    def _verify_command(self) -> None:
+        if not self.command:
+            self.generate_command()
+        if not self.command:
+            msg = 'Flash or run command is empty, please verify if it was generated properly.'
+            logger.error(msg)
+            raise TwisterHarnessException(msg)
+
+    def _start_reader_thread(self) -> None:
+        self._reader_thread = threading.Thread(target=self._handle_device_output, daemon=True)
+        self._reader_thread.start()
+
+    def _handle_device_output(self) -> None:
+        """
+        This method is dedicated to run it in separate thread to read output
+        from device and put them into internal queue and save to log file.
+        At the end of reading put END_OF_DATA object to _device_read_queue, to
+        indicate other threads that there will be no more data to read
+        """
+        with open(self.handler_log_path, 'a+') as log_file:
+            while self._is_device_running():
+                if self._is_device_connected():
+                    output = self._read_device_output().decode().strip()
+                    if output:
+                        self._device_read_queue.put(output)
+                        log_file.write(f'{output}\n')
+                else:
+                    # ignore output from device
+                    self._flush_device_output()
+                    time.sleep(0.1)
+        # TODO: what should be END_OF_DATA instance of object, or rather string
```

Comment:
Uprościłoby to annotacje w kilku metodach (`str | object`) ...
Ja się zastanawiam, czy potrzebujemy ten obiekt (choć lepiej żeby był, bo nie chcę później rozwiązywać problemów z synchronizacją ... bo już miesza mi się kiedy używamy `is_device_connected` a kiedy `is_device_running`)

Reply:
ok, sam sobie odpowiem: `END_OF_DATA` jest tylko o to, żeby pozwolić odczytać kolejkę, nawet jak już device is not connected